File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/08_linked_lists/MergeTwoSortedLists.cs
// LeetCode 21 - Merge Two Sorted Lists
// Difficulty: Easy
// Pattern: Dummy head + two pointers
//
// Time: O(m + n)  Space: O(1)

namespace CodingPatterns.LinkedLists;

public class ListNode
{
    public int Val;
    public ListNode Next;
    public ListNode(int val = 0, ListNode next = null) { Val = val; Next = next; }
}

public class MergeTwoSortedLists
{
    public ListNode MergeLists(ListNode list1, ListNode list2)
    {
        var dummy = new ListNode(0);
        var curr = dummy;

        while (list1 != null && list2 != null)
        {
            if (list1.Val <= list2.Val)
            { curr.Next = list1; list1 = list1.Next; }
            else
            { curr.Next = list2; list2 = list2.Next; }
            curr = curr.Next;
        }
        curr.Next = list1 ?? list2;
        return dummy.Next;
    }

    // ---- Tests ----
    public static void Run()
    {
        var sol = new MergeTwoSortedLists();

        var a = new ListNode(1, new ListNode(2, new ListNode(4)));
        var b = new ListNode(1, new ListNode(3, new ListNode(4)));
        Console.WriteLine(Format(sol.MergeLists(a, b)));            // 1,1,2,3,4,4

        Console.WriteLine(Format(sol.MergeLists(null, null)));      // (empty)
        Console.WriteLine(Format(sol.MergeLists(null, new ListNode(0))));  // 0
    }

    private static string Format(ListNode head)
    {
        var vals = new List<int>();
        for (var node = head; node != null; node = node.Next)
            vals.Add(node.Val);
        return string.Join(",", vals);
    }
}

File: Coding-Patterns/Coding-Patterns-Python/08_linked_lists/merge_two_sorted_lists.py
# LeetCode 21 - Merge Two Sorted Lists
# Difficulty: Easy
# Pattern: Dummy head + two pointers
#
# Problem: Merge two sorted linked lists and return the merged list.
#
# Approach: Use a dummy head to simplify building the result list.
# Advance the pointer that points to the smaller value.
# Attach the remaining non-null list at the end.
#
# Time: O(m + n)  Space: O(1)

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def merge_two_lists(list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    dummy = ListNode(0)
    curr = dummy

    while list1 and list2:
        if list1.val <= list2.val:
            curr.next = list1
            list1 = list1.next
        else:
            curr.next = list2
            list2 = list2.next
        curr = curr.next

    curr.next = list1 or list2  # attach remainder
    return dummy.next


# ---- Helpers ----
def to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result

def from_list(vals):
    dummy = ListNode(0)
    curr = dummy
    for v in vals:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


# ---- Tests ----
if __name__ == "__main__":
    l1 = from_list([1, 2, 4])
    l2 = from_list([1, 3, 4])
    assert to_list(merge_two_lists(l1, l2)) == [1, 1, 2, 3, 4, 4]
    assert to_list(merge_two_lists(None, None)) == []
    assert to_list(merge_two_lists(None, from_list([0]))) == [0]
    print("All tests passed.")
